mmap.h

```diff
@@ -215,11 +240,52 @@ static void *WinMap(void *addr, size_t length, int prot,
                              (offset + length) >> 32,
                              (offset + length), 0);
     if (!hand) {
+        LPVOID lpMsgBuf;
+        LPVOID lpDisplayBuf;
+        DWORD error_code = GetLastError();
+        FormatMessage(
+            FORMAT_MESSAGE_ALLOCATE_BUFFER |
+            FORMAT_MESSAGE_FROM_SYSTEM |
+            FORMAT_MESSAGE_IGNORE_INSERTS,
+            NULL,
+            error_code,
+            MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+            (LPTSTR)&lpMsgBuf,
+            0, NULL);
+        lpDisplayBuf = (LPVOID)LocalAlloc(LMEM_ZEROINIT,
+            (lstrlen((LPCTSTR)lpMsgBuf) + 256) * sizeof(TCHAR));
+        StringCchPrintf((LPTSTR)lpDisplayBuf,
+            LocalSize(lpDisplayBuf) / sizeof(TCHAR),
+            TEXT("%s failed with error %d: %s"),
+            error_code, lpMsgBuf);
         return MAP_FAILED;
     }
+    if (winprot == PAGE_WRITECOPY) {
+        access = FILE_MAP_COPY;
+    }
+
     res = MapViewOfFileEx(hand, access, offset >> 32,
                           offset, length, addr);
     if (!res) {
+        LPVOID lpMsgBuf;
+        LPVOID lpDisplayBuf;
+        DWORD error_code = GetLastError();
+        FormatMessage(
+            FORMAT_MESSAGE_ALLOCATE_BUFFER |
+            FORMAT_MESSAGE_FROM_SYSTEM |
+            FORMAT_MESSAGE_IGNORE_INSERTS,
+            NULL,
+            error_code,
+            MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+            (LPTSTR)&lpMsgBuf,
+            0, NULL);
+        lpDisplayBuf = (LPVOID)LocalAlloc(LMEM_ZEROINIT,
+            (lstrlen((LPCTSTR)lpMsgBuf) + 40) * sizeof(TCHAR));
+        StringCchPrintf((LPTSTR)lpDisplayBuf,
+            LocalSize(lpDisplayBuf) / sizeof(TCHAR),
+            TEXT("failed with error %d: %s"),
+            error_code, lpMsgBuf);
+        fprintf(stderr, (char*)lpDisplayBuf);
```

Comment:
W.r.t. fprintf, I'd say if you're already going to the trouble of pulling out all of the above WIN32 functions, then just go for the gold and use `WriteFile(GetStdHandle(STD_ERROR_HANDLE))`. 

[side commentary] In my work on Cosmopolitan Libc, one thing I love doing for instance, is ensuring that Cosmo *never* depends on the MSVC Libc, since it has a history of bundling things like telemetry, plus there's like 10 different Microsoft Libc's to choose from. But linking just KERNEL32 is awesome when it's possible, which is what Cosmo apps do.